File: tests/wasm/test_dma.js
/**
 * WASM binding tests for DMA (Dickson Moving Average) indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module:', error);
        throw error;
    }
    
    
    testData = loadTestData();
});

test('DMA partial params', () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.dma;
    
    const result = wasm.dma_js(
        close,
        expected.defaultParams.hull_length,
        expected.defaultParams.ema_length,
        expected.defaultParams.ema_gain_limit,
        expected.defaultParams.hull_ma_type
    );
    assert.strictEqual(result.length, close.length);
    
    
    const hullLen = expected.defaultParams.hull_length;
    const emaLen = expected.defaultParams.ema_length;
    const warmup = Math.max(hullLen, emaLen) - 1;
    
    assertAllNaN(result.slice(0, warmup), `Expected NaN during warmup period (first ${warmup} values)`);
    
    
    let nonNanCount = 0;
    for (let i = warmup; i < result.length; i++) {
        if (!isNaN(result[i])) nonNanCount++;
    }
    assert(nonNanCount > close.length - warmup - 10, "Should have values after warmup");
});

test('DMA accuracy', () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.dma;
    
    const result = wasm.dma_js(
        close,
        expected.defaultParams.hull_length,
        expected.defaultParams.ema_length,
        expected.defaultParams.ema_gain_limit,
        expected.defaultParams.hull_ma_type
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        0.001,  
        "DMA last 5 values mismatch"
    );
});

test('DMA default candles', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.dma_js(close, 7, 20, 50, "WMA");
    assert.strictEqual(result.length, close.length);
    
    
    const warmup = Math.max(7, 20) - 1;
    
    assertAllNaN(result.slice(0, warmup), `Expected NaN in first ${warmup} values`);
    
    
    let hasValues = false;
    for (let i = warmup; i < result.length; i++) {
        if (!isNaN(result[i])) {
            hasValues = true;
            break;
        }
    }
    assert(hasValues, "Should have values after warmup");
});

test('DMA zero hull period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.dma_js(inputData, 0, 20, 50, "WMA");
    }, /Invalid period/);
});

test('DMA zero ema period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.dma_js(inputData, 7, 0, 50, "WMA");
    }, /Invalid period/);
});

test('DMA empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.dma_js(empty, 7, 20, 50, "WMA");
    }, /empty/i);
});

test('DMA all NaN', () => {
    
    const nanData = new Float64Array([NaN, NaN, NaN]);
    
    assert.throws(() => {
        wasm.dma_js(nanData, 7, 20, 50, "WMA");
    }, /NaN/i);
});

test('DMA invalid hull type', () => {
    
    const inputData = new Float64Array(Array(50).fill(10.0));
    
    assert.throws(() => {
        wasm.dma_js(inputData, 7, 20, 50, "INVALID");
    }, /Invalid Hull MA type/);
});

test('DMA EMA hull type', () => {
    
    const inputData = new Float64Array(Array.from({length: 100}, (_, i) => i));
    
    const result = wasm.dma_js(
        inputData,
        7,    
        20,   
        50,   
        "EMA" 
    );
    
    assert.strictEqual(result.length, inputData.length);
    
    const nonNanCount = result.filter(x => !isNaN(x)).length;
    assert(nonNanCount > 0, 'Should produce non-NaN values after warmup');
});

test('DMA period exceeds length', () => {
    
    const smallData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.dma_js(smallData, 10, 20, 50, "WMA");
    }, /Invalid period/);
});

test('DMA insufficient data', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.dma_js(singlePoint, 7, 20, 50, "WMA");
    });
});

test('DMA NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.dma;
    
    const result = wasm.dma_js(
        close,
        expected.defaultParams.hull_length,
        expected.defaultParams.ema_length,
        expected.defaultParams.ema_gain_limit,
        expected.defaultParams.hull_ma_type
    );
    
    assert.strictEqual(result.length, close.length);
    
    
    const hullLen = expected.defaultParams.hull_length;
    const emaLen = expected.defaultParams.ema_length;
    const warmup = Math.max(hullLen, emaLen) - 1;
    
    
    if (result.length > warmup + 100) {
        
        for (let i = warmup + 100; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
    
    
    assertAllNaN(result.slice(0, warmup), `Expected NaN in warmup period (first ${warmup} values)`);
});




test('DMA constant input', () => {
    
    const constantVal = EXPECTED_OUTPUTS.dma.constantValue;
    const inputData = new Float64Array(100).fill(constantVal);
    
    const result = wasm.dma_js(inputData, 7, 20, 50, "WMA");
    
    assert.strictEqual(result.length, inputData.length);
    
    
    
    const last10 = result.slice(-10);
    const validValues = last10.filter(x => !isNaN(x));
    
    if (validValues.length > 0) {
        
        validValues.forEach((val, i) => {
            assertClose(
                val, 
                constantVal,
                0.01,  
                `DMA should converge to constant input value at index ${i}`
            );
        });
    }
});

test('DMA mixed hull types', () => {
    
    const inputData = new Float64Array(Array.from({length: 100}, (_, i) => i * 10 + Math.sin(i)));
    const expected = EXPECTED_OUTPUTS.dma;
    
    const results = {};
    for (const hullType of expected.hullMaTypes) {
        const result = wasm.dma_js(
            inputData,
            7,    
            20,   
            50,   
            hullType
        );
        
        assert.strictEqual(result.length, inputData.length, `Result length mismatch for hull_type=${hullType}`);
        
        
        const nonNanCount = result.filter(x => !isNaN(x)).length;
        assert(nonNanCount > 0, `Should produce values for hull_type=${hullType}`);
        
        results[hullType] = result;
    }
    
    
    if (results['EMA'] && results['WMA']) {
        let hasDifference = false;
        for (let i = 0; i < results['EMA'].length; i++) {
            if (!isNaN(results['EMA'][i]) && !isNaN(results['WMA'][i])) {
                if (Math.abs(results['EMA'][i] - results['WMA'][i]) > 1e-10) {
                    hasDifference = true;
                    break;
                }
            }
        }
        assert(hasDifference, "EMA and WMA hull types should produce different results");
    }
});

test('DMA gain limit edge cases', () => {
    
    const inputData = new Float64Array(Array.from({length: 50}, (_, i) => i));
    
    
    const resultZero = wasm.dma_js(
        inputData,
        7,    
        20,   
        0,    
        "WMA"
    );
    assert.strictEqual(resultZero.length, inputData.length, "Zero gain limit should produce output");
    
    
    const resultLarge = wasm.dma_js(
        inputData,
        7,     
        20,    
        1000,  
        "WMA"
    );
    assert.strictEqual(resultLarge.length, inputData.length, "Large gain limit should produce output");
});

test('DMA trending data', () => {
    
    
    const uptrend = new Float64Array(Array.from({length: 100}, (_, i) => i));
    const resultUp = wasm.dma_js(uptrend, 7, 20, 50, "WMA");
    assert.strictEqual(resultUp.length, uptrend.length);
    
    
    const downtrend = new Float64Array(Array.from({length: 100}, (_, i) => 100 - i));
    const resultDown = wasm.dma_js(downtrend, 7, 20, 50, "WMA");
    assert.strictEqual(resultDown.length, downtrend.length);
    
    
    const warmup = 19; 
    
    
    if (resultUp.length > warmup + 10) {
        const last10 = resultUp.slice(-10);
        const valid = last10.filter(x => !isNaN(x));
        if (valid.length > 1) {
            
            let increasingCount = 0;
            for (let i = 1; i < valid.length; i++) {
                if (valid[i] > valid[i-1]) increasingCount++;
            }
            assert(increasingCount > valid.length / 2, "DMA should follow uptrend");
        }
    }
});

test('DMA oscillating data', () => {
    
    
    const points = 100;
    const oscillating = new Float64Array(Array.from({length: points}, (_, i) => {
        const x = (i / points) * 4 * Math.PI;
        return 50.0 + 10.0 * Math.sin(x);
    }));
    
    const result = wasm.dma_js(
        oscillating,
        7,    
        20,   
        50,   
        "WMA"
    );
    
    assert.strictEqual(result.length, oscillating.length);
    
    
    const warmup = 19;
    if (result.length > warmup) {
        const validResult = result.slice(warmup).filter(x => !isNaN(x));
        const validInput = oscillating.slice(warmup);
        
        
        const mean = arr => arr.reduce((a, b) => a + b, 0) / arr.length;
        const variance = arr => {
            const m = mean(arr);
            return arr.reduce((a, b) => a + Math.pow(b - m, 2), 0) / arr.length;
        };
        
        if (validResult.length > 0) {
            const resultVar = variance(validResult);
            const inputVar = variance(validInput);
            
            
            assert(resultVar < inputVar, "DMA should smooth oscillating data");
        }
    }
});

test('DMA extreme ratios', () => {
    
    const inputData = new Float64Array(Array.from({length: 100}, (_, i) => i * 2 + Math.random()));
    
    
    const result1 = wasm.dma_js(
        inputData,
        3,    
        50,   
        50,   
        "WMA"
    );
    assert.strictEqual(result1.length, inputData.length);
    
    
    const result2 = wasm.dma_js(
        inputData,
        50,   
        3,    
        50,   
        "WMA"
    );
    assert.strictEqual(result2.length, inputData.length);
    
    
    let hasDifference = false;
    for (let i = 0; i < result1.length; i++) {
        if (!isNaN(result1[i]) && !isNaN(result2[i])) {
            if (Math.abs(result1[i] - result2[i]) > 0.01) {
                hasDifference = true;
                break;
            }
        }
    }
    assert(hasDifference, "Different hull/ema ratios should produce different results");
});

test('DMA zero-copy API', () => {
    
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]);
    const hullLength = 5;
    const emaLength = 10;
    const emaGainLimit = 50;
    const hullMaType = "WMA";
    
    
    const ptr = wasm.dma_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');
    
    
    const memoryBuffer = wasm.__wasm.memory.buffer;
    const memView = new Float64Array(
        memoryBuffer,
        ptr,
        data.length
    );
    
    
    memView.set(data);
    
    try {
        
        wasm.dma_into(
            ptr,
            ptr,
            data.length,
            hullLength,
            emaLength,
            emaGainLimit,
            hullMaType
        );
        
        
        const regularResult = wasm.dma_js(data, hullLength, emaLength, emaGainLimit, hullMaType);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue; 
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {
        
        wasm.dma_free(ptr, data.length);
    }
});

test('DMA zero-copy error handling', () => {
    
    assert.throws(() => {
        wasm.dma_into(0, 0, 10, 7, 20, 50, "WMA");
    }, /null pointer|invalid memory/i);
    
    
    const ptr = wasm.dma_alloc(10);
    try {
        
        assert.throws(() => {
            wasm.dma_into(ptr, ptr, 10, 0, 20, 50, "WMA");
        }, /Invalid period/);
        
        
        assert.throws(() => {
            wasm.dma_into(ptr, ptr, 10, 7, 0, 50, "WMA");
        }, /Invalid period/);
        
        
        
        assert.throws(() => {
            wasm.dma_into(ptr, ptr, 10, 7, 20, 50, "INVALID");
        });
    } finally {
        wasm.dma_free(ptr, 10);
    }
});

test('DMA memory management', () => {
    
    const sizes = [100, 1000, 10000];
    
    for (const size of sizes) {
        const ptr = wasm.dma_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);
        
        
        const memoryBuffer = wasm.__wasm.memory.buffer;
        const memView = new Float64Array(memoryBuffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }
        
        
        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }
        
        
        wasm.dma_free(ptr, size);
    }
});

test.after(() => {
    console.log('DMA WASM tests completed');
});

File: tests/python/test_dma.py
"""
Python binding tests for DMA (Dickson Moving Average) indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
import os
from pathlib import Path


sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

try:
    import my_project as ta_indicators
except ImportError:
    pytest.skip("Python module not built", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS


class TestDma:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_dma_partial_params(self, test_data):
        """Test DMA with default parameters - mirrors check_dma_partial_params"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['dma']
        
        
        result = ta_indicators.dma(
            close,
            hull_length=expected['default_params']['hull_length'],
            ema_length=expected['default_params']['ema_length'],
            ema_gain_limit=expected['default_params']['ema_gain_limit'],
            hull_ma_type=expected['default_params']['hull_ma_type']
        )
        assert len(result) == len(close)
        
        
        hull_len = expected['default_params']['hull_length']
        ema_len = expected['default_params']['ema_length']
        warmup = max(hull_len, ema_len) - 1
        
        assert np.all(np.isnan(result[:warmup])), f"Expected NaN during warmup period (first {warmup} values)"
        assert np.sum(~np.isnan(result)) > len(close) - warmup - 10, "Should have values after warmup"
    
    def test_dma_accuracy(self, test_data):
        """Test DMA matches expected values from Rust tests - mirrors check_dma_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['dma']
        
        result = ta_indicators.dma(
            close,
            hull_length=expected['default_params']['hull_length'],
            ema_length=expected['default_params']['ema_length'],
            ema_gain_limit=expected['default_params']['ema_gain_limit'],
            hull_ma_type=expected['default_params']['hull_ma_type']
        )
        
        assert len(result) == len(close)
        
        
        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=0.001,  
            atol=0.001,
            msg="DMA last 5 values mismatch"
        )
    
    def test_dma_default_candles(self, test_data):
        """Test DMA with default parameters - mirrors check_dma_default_candles"""
        close = test_data['close']
        
        
        result = ta_indicators.dma(close, 7, 20, 50, "WMA")
        assert len(result) == len(close)
        
        
        warmup = max(7, 20) - 1
        
        assert np.all(np.isnan(result[:warmup])), f"Expected NaN in first {warmup} values"
        assert not np.all(np.isnan(result[warmup:])), "Should have values after warmup"
    
    def test_dma_zero_hull_period(self):
        """Test DMA fails with zero hull period - mirrors check_dma_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.dma(input_data, hull_length=0, ema_length=20, ema_gain_limit=50, hull_ma_type="WMA")
    
    def test_dma_zero_ema_period(self):
        """Test DMA fails with zero EMA period - mirrors check_dma_zero_ema_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.dma(input_data, hull_length=7, ema_length=0, ema_gain_limit=50, hull_ma_type="WMA")
    
    def test_dma_period_exceeds_length(self):
        """Test DMA fails when period exceeds data length - mirrors check_dma_period_exceeds_length"""
        small_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.dma(small_data, hull_length=10, ema_length=20, ema_gain_limit=50, hull_ma_type="WMA")
    
    def test_dma_very_small_dataset(self):
        """Test DMA fails with insufficient data - mirrors check_dma_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError):
            ta_indicators.dma(single_point, hull_length=7, ema_length=20, ema_gain_limit=50, hull_ma_type="WMA")
    
    def test_dma_empty_input(self):
        """Test DMA fails with empty input - mirrors check_dma_empty_input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.dma(empty, hull_length=7, ema_length=20, ema_gain_limit=50, hull_ma_type="WMA")
    
    def test_dma_all_nan(self):
        """Test DMA fails with all NaN values - mirrors check_dma_all_nan"""
        nan_data = np.array([np.nan, np.nan, np.nan])
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.dma(nan_data, hull_length=7, ema_length=20, ema_gain_limit=50, hull_ma_type="WMA")
    
    def test_dma_invalid_hull_type(self):
        """Test DMA fails with invalid hull_ma_type - mirrors check_dma_invalid_hull_type"""
        input_data = np.array([10.0] * 50)
        
        with pytest.raises(ValueError, match="Invalid Hull MA type"):
            ta_indicators.dma(
                input_data,
                hull_length=7,
                ema_length=20,
                ema_gain_limit=50,
                hull_ma_type="INVALID"
            )
    
    def test_dma_ema_hull_type(self):
        """Test DMA works with EMA hull type - mirrors check_dma_ema_hull_type"""
        input_data = np.array([float(i) for i in range(100)])
        
        result = ta_indicators.dma(
            input_data,
            hull_length=7,
            ema_length=20,
            ema_gain_limit=50,
            hull_ma_type="EMA"
        )
        
        assert len(result) == len(input_data)
        
        assert np.sum(~np.isnan(result)) > 0, "Should produce non-NaN values after warmup"
    
    def test_dma_nan_handling(self, test_data):
        """Test DMA handles NaN values correctly - mirrors check_dma_nan_handling"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['dma']
        
        result = ta_indicators.dma(
            close,
            hull_length=expected['default_params']['hull_length'],
            ema_length=expected['default_params']['ema_length'],
            ema_gain_limit=expected['default_params']['ema_gain_limit'],
            hull_ma_type=expected['default_params']['hull_ma_type']
        )
        
        assert len(result) == len(close)
        
        
        hull_len = expected['default_params']['hull_length']
        ema_len = expected['default_params']['ema_length']
        warmup = max(hull_len, ema_len) - 1
        
        
        if len(result) > warmup + 100:
            
            assert not np.any(np.isnan(result[warmup+100:])), "Found unexpected NaN after warmup period"
        
        
        assert np.all(np.isnan(result[:warmup])), f"Expected NaN in warmup period (first {warmup} values)"
    
    
    
    
    def test_dma_batch_default_params(self, test_data):
        """Test DMA batch processing with default parameters only"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['dma']
        
        result = ta_indicators.dma_batch(
            close,
            hull_length_range=(7, 7, 0),  
            ema_length_range=expected['batch_ema_range'],      
            ema_gain_limit_range=expected['batch_gain_range'], 
            hull_ma_type="WMA"
        )
        
        assert 'values' in result
        assert 'hull_lengths' in result
        assert 'ema_lengths' in result
        assert 'ema_gain_limits' in result
        
        
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)
        
        
        default_row = result['values'][0]
        
        
        assert_close(
            default_row[-5:],
            expected['batch_default_row'],
            rtol=0.001,
            atol=0.001,
            msg="DMA batch default row mismatch"
        )
    
    def test_dma_batch_hull_sweep(self, test_data):
        """Test DMA batch processing with hull length parameter sweep"""
        close = test_data['close'][:500]  
        expected = EXPECTED_OUTPUTS['dma']
        
        result = ta_indicators.dma_batch(
            close,
            hull_length_range=expected['batch_hull_range'],  
            ema_length_range=(20, 20, 0),  
            ema_gain_limit_range=(50, 50, 0),  
            hull_ma_type="WMA"
        )
        
        
        assert result['values'].shape[0] == 4
        assert result['values'].shape[1] == len(close)
        assert len(result['hull_lengths']) == 4
        assert list(result['hull_lengths']) == expected['batch_hull_lengths']
        
        
        for i, hull_len in enumerate(result['hull_lengths']):
            row = result['values'][i]
            
            warmup = max(hull_len, 20) - 1
            
            
            assert np.all(np.isnan(row[:warmup])), f"Expected NaN in warmup for hull_length={hull_len}"
            
            
            if len(row) > warmup + 10:
                non_nan_count = np.sum(~np.isnan(row[warmup:]))
                assert non_nan_count > 0, f"Should have values after warmup for hull_length={hull_len}"
    
    def test_dma_batch_full_sweep(self, test_data):
        """Test DMA batch with multiple parameter combinations"""
        close = test_data['close'][:100]  
        
        result = ta_indicators.dma_batch(
            close,
            hull_length_range=(5, 7, 2),      
            ema_length_range=(15, 20, 5),     
            ema_gain_limit_range=(40, 50, 10), 
            hull_ma_type="WMA"
        )
        
        
        assert result['values'].shape[0] == 8
        assert result['values'].shape[1] == len(close)
        
        
        assert len(result['hull_lengths']) == 8
        assert len(result['ema_lengths']) == 8
        assert len(result['ema_gain_limits']) == 8
        
        
        assert result['hull_lengths'][0] == 5
        assert result['ema_lengths'][0] == 15
        assert result['ema_gain_limits'][0] == 40
        
        assert result['hull_lengths'][-1] == 7
        assert result['ema_lengths'][-1] == 20
        assert result['ema_gain_limits'][-1] == 50
    
    def test_dma_constant_input(self):
        """Test DMA with constant input values"""
        constant_val = EXPECTED_OUTPUTS['dma']['constant_value']
        input_data = np.full(100, constant_val)
        
        result = ta_indicators.dma(input_data, 7, 20, 50, "WMA")
        
        assert len(result) == len(input_data)
        
        
        
        last_10 = result[-10:]
        valid_values = last_10[~np.isnan(last_10)]
        
        if len(valid_values) > 0:
            
            assert_close(
                valid_values,
                np.full(len(valid_values), constant_val),
                rtol=0.01,  
                msg="DMA should converge to constant input value"
            )
    
    
    
    
    def test_dma_batch_edge_cases(self):
        """Test DMA batch processing edge cases"""
        close = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0] * 5)  
        
        
        single_batch = ta_indicators.dma_batch(
            close,
            hull_length_range=(5, 5, 0),
            ema_length_range=(10, 10, 0),
            ema_gain_limit_range=(30, 30, 0),
            hull_ma_type="WMA"
        )
        
        assert single_batch['values'].shape[0] == 1
        assert len(single_batch['hull_lengths']) == 1
        
        
        large_step = ta_indicators.dma_batch(
            close,
            hull_length_range=(5, 7, 10),  
            ema_length_range=(10, 10, 0),
            ema_gain_limit_range=(30, 30, 0),
            hull_ma_type="WMA"
        )
        
        
        assert large_step['values'].shape[0] == 1
        assert large_step['hull_lengths'][0] == 5
    
    def test_dma_mixed_hull_types(self):
        """Test DMA with different Hull MA types"""
        input_data = np.array([float(i * 10 + np.sin(i)) for i in range(100)])
        results = {}
        
        for hull_type in EXPECTED_OUTPUTS['dma']['hull_ma_types']:
            result = ta_indicators.dma(
                input_data,
                hull_length=7,
                ema_length=20,
                ema_gain_limit=50,
                hull_ma_type=hull_type
            )
            
            assert len(result) == len(input_data), f"Result length mismatch for hull_type={hull_type}"
            
            
            non_nan_count = np.sum(~np.isnan(result))
            assert non_nan_count > 0, f"Should produce values for hull_type={hull_type}"
            
            results[hull_type] = result
        
        
        if 'EMA' in results and 'WMA' in results:
            ema_result = results['EMA']
            wma_result = results['WMA']
            valid_idx = ~(np.isnan(ema_result) | np.isnan(wma_result))
            if np.sum(valid_idx) > 0:
                assert not np.allclose(ema_result[valid_idx], wma_result[valid_idx], rtol=1e-10), \
                    "EMA and WMA hull types should produce different results"
    
    def test_dma_gain_limit_edge_cases(self):
        """Test DMA with edge case gain limit values"""
        input_data = np.array([float(i) for i in range(50)])
        
        
        result_zero = ta_indicators.dma(
            input_data,
            hull_length=7,
            ema_length=20,
            ema_gain_limit=0,
            hull_ma_type="WMA"
        )
        assert len(result_zero) == len(input_data), "Zero gain limit should produce output"
        
        
        result_large = ta_indicators.dma(
            input_data,
            hull_length=7,
            ema_length=20,
            ema_gain_limit=1000,
            hull_ma_type="WMA"
        )
        assert len(result_large) == len(input_data), "Large gain limit should produce output"
    
    def test_dma_trending_data(self):
        """Test DMA with perfectly trending data"""
        
        uptrend = np.array([float(i) for i in range(100)])
        result_up = ta_indicators.dma(uptrend, 7, 20, 50, "WMA")
        assert len(result_up) == len(uptrend)
        
        
        downtrend = np.array([100.0 - float(i) for i in range(100)])
        result_down = ta_indicators.dma(downtrend, 7, 20, 50, "WMA")
        assert len(result_down) == len(downtrend)
        
        
        warmup = 19  
        
        
        if len(result_up) > warmup + 10:
            last_10 = result_up[-10:]
            valid = last_10[~np.isnan(last_10)]
            if len(valid) > 1:
                
                diffs = np.diff(valid)
                assert np.mean(diffs) > 0, "DMA should follow uptrend"
    
    def test_dma_oscillating_data(self):
        """Test DMA with oscillating/choppy data"""
        
        x = np.linspace(0, 4 * np.pi, 100)
        oscillating = 50.0 + 10.0 * np.sin(x)
        
        result = ta_indicators.dma(
            oscillating,
            hull_length=7,
            ema_length=20,
            ema_gain_limit=50,
            hull_ma_type="WMA"
        )
        
        assert len(result) == len(oscillating)
        
        
        warmup = 19
        if len(result) > warmup:
            valid_result = result[warmup:]
            valid_input = oscillating[warmup:]
            
            
            result_var = np.nanvar(valid_result)
            input_var = np.var(valid_input)
            
            
            assert result_var < input_var, "DMA should smooth oscillating data"
    
    def test_dma_intermittent_nan(self):
        """Test DMA with data containing intermittent NaN values"""
        data = np.array([float(i) for i in range(100)])
        
        data[25] = np.nan
        data[50] = np.nan
        data[75] = np.nan
        
        
        result = ta_indicators.dma(data, 7, 20, 50, "WMA")
        assert len(result) == len(data)
        
        
        non_nan_count = np.sum(~np.isnan(result))
        assert non_nan_count > 0, "Should produce some non-NaN values"
    
    def test_dma_extreme_ratios(self):
        """Test DMA with extreme hull/ema length ratios"""
        input_data = np.array([float(i * 2 + np.random.randn()) for i in range(100)])
        
        
        result1 = ta_indicators.dma(
            input_data,
            hull_length=3,
            ema_length=50,
            ema_gain_limit=50,
            hull_ma_type="WMA"
        )
        assert len(result1) == len(input_data)
        
        
        result2 = ta_indicators.dma(
            input_data,
            hull_length=50,
            ema_length=3,
            ema_gain_limit=50,
            hull_ma_type="WMA"
        )
        assert len(result2) == len(input_data)
        
        
        valid_idx = ~(np.isnan(result1) | np.isnan(result2))
        if np.sum(valid_idx) > 0:
            assert not np.allclose(result1[valid_idx], result2[valid_idx], rtol=0.01), \
                "Different hull/ema ratios should produce different results"


if __name__ == "__main__":
    pytest.main([__file__, "-v"])